Export and Import (#243)
javascript-tutorial/it.javascript.info: 1-js/13-modules/02-import-export/article.md

# Export e Import

Le direttive export e import esistono sotto forma di diverse varianti.

Nell'articolo precedente, ne abbiamo visto un utilizzo molto semplice, ora esploriamo più esempi.

## Export prima della dichiarazione

Possiamo etichettare una qualsiasi dichiarazione come esportata, con il prefisso `export`, sia che questa sia una variabile, una funzione o una classe.

Ad esempio, i seguenti exports sono tutti validi:

```js
// esportiamo un array
*!*export*/!* let months = ['Jan', 'Feb', 'Mar','Apr', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];

// esportiamo una costante
*!*export*/!* const MODULES_BECAME_STANDARD_YEAR = 2015;

// esportiamo una classe
*!*export*/!* class User {
  constructor(name) {
    this.name = name;
  }
}
```

````smart header="Non c'è bisogno del punto e virgola dopo l'export di una classe/funzione"
Da notare che il termine `export` prima di una funzione, non la rende un'[espressione di funzione](info:function-expressions). Rimane sempre una dichiarazione di funzione, viene semplicemente esportata.

> Molte style guide di JavaScript sconsigliano l'utilizzo del punto e virgola dopo la dichiarazione di una classe o funzione.

Reviewer: ```suggestion
Molte guide style di JavaScript sconsigliano l'utilizzo del punto e virgola dopo la dichiarazione di una classe o funzione.
```

> Molte guide style di JavaScript sconsigliano l'utilizzo del punto e virgola dopo la dichiarazione di una classe o funzione.

Reviewer 2: ```suggestion
Molte *style guides* di JavaScript sconsigliano l'utilizzo del punto e virgola dopo la dichiarazione di una classe o funzione.
```

sayBye('John'); // Bye, John!
```

> Ma ne caso in cui si fossero molti import, potremmo importare tutto come un oggetto, utilizzando `import * as <obj>`, ad esempio:

Reviewer: ```suggestion
Ma nel caso in cui ci fossero molti import, potremmo importare tutto come un oggetto, utilizzando `import * as <obj>`, ad esempio:
```

## Export default

> Nella pratica, esistono principalmente due tipi di moduli.

Reviewer: ```suggestion
In pratica, esistono principalmente due tipi di moduli.
```

> 1. Moduli che contengono una library, pacchetti di funzioni, come `say.js` visto sopra.

Reviewer 2: ```suggestion
1. Moduli che contengono una libreria: pacchetti di funzioni, come `say.js` visto sopra.
```

> 2. Moduli che dichiarano una singola entità, esempio un modulo `user.js` che esporta solamente `class User`.

Reviewer 2: ```suggestion
2. Moduli che dichiarano una singola entità: esempio un modulo `user.js` che esporta solamente `class User`.
```

Nella maggior parte dei casi, si preferisce il secondo approccio, in modo tale che "ogni cosa" stia nel suo modulo.

> Naturalmente, questa pratica richiede l'utilizzo di molti files, poiché ogni cosa richiede il suo modulo, ma questo non è un problema. In realtà, la navigazione del codice diventa più semplice se tutti i file hanno dei nomi parlanti e sono ben strutturati all'interno di cartelle.

Reviewer: ```suggestion
Naturalmente, questa pratica richiede l'utilizzo di molti files, poiché ogni cosa richiede il suo modulo, ma questo non è un problema. In realtà, la navigazione del codice diventa più semplice se tutti i file hanno dei nomi descrittivi e sono ben strutturati all'interno di cartelle.
```

> I moduli forniscono una speciali sintassi, `export default` ("export di default"), per rendere la pratica "una cosa per modulo" più elegante.

Reviewer: ```suggestion
I moduli forniscono una speciale sintassi, `export default` ("export di default"), per rendere la pratica "una cosa per modulo" più elegante.
```

> E' sufficiente inserire `export default`, prima dell'entità da esportare:

Reviewer: ```suggestion
E' sufficiente inserire `export default` prima dell'entità da esportare:
```

```js
// 📁 user.js
export *!*default*/!* class User { // aggiungiamo "default"
  constructor(name) {
    this.name = name;
  }
}
```

Può esserci solamente un `export default` per file.

...E possiamo importarlo senza le parentesi graffe:

```js
// 📁 main.js
import *!*User*/!* from './user.js'; // non {User}, semplicemente User

new User('John');
```

Gli import senza parentesi graffe sono più eleganti. Un errore comune quando si inizia ad utilizzare i moduli è quello di dimenticarsi le parentesi graffe. Quindi, ricorda, `import` richiede le parentesi graffe per i named export, ma non le richiedere per i default export.

| Named export | Default export |
|--------------|----------------|
| `export class User {...}` | `export default class User {...}` |
| `import {User} from ...` | `import User from ...`|

> Tecnicamente, potremmo avere sia il default che il named export nello stesso modulo, ma nella pratica gli sviluppatori non lo fanno. Un modulo può essere named export o default export.

Reviewer: ```suggestion
Tecnicamente, potremmo avere sia il default che il named export nello stesso modulo, ma in pratica gli sviluppatori non lo fanno. Un modulo può essere named export o default export.
```

new User('John');
```

E, infine, se importiamo tutto `*` come un oggetto, allora la proprietà di `default` corrisponde al default export:

```js
// 📁 main.js
import * as user from './user.js';

let User = user.default; // il default export
new User('John');
```

### Una parola contro il default exports

> I named export sono espliciti. Elencano esattamente il nome di ciò che vogliono importare, quindi avremo questa informazione (il nome) da loro; questa è sempre una buona cosa.

Reviewer 2: ```suggestion
I named export sono espliciti, elencano esattamente il nome di ciò che vogliono importare. Avere in chiaro questa informazione (il nome), è sempre una buona cosa.
```

I named export ci forzano ad utilizzare il nome esatto di ciò che vogliamo importare:

```js
import {User} from './user.js';
// import {MyUser} non funzionerebbe, il nome deve essere {User}
```

> ...Mentre per un default export, noi possiamo decidere il nome in fase di importazione:

Reviewer: ```suggestion
...Mentre per un default export, possiamo decidere il nome in fase di importazione:
```

```js
import User from './user.js'; // funziona
import MyUser from './user.js'; // funziona

> // potrebbe essere anche import Anything... e comunque funzionerebbe

Reviewer: ```suggestion
// potrebbe essere anche import Anything... e funzionerebbe comunque
```

```

> Quindi i membri del team potrebbero utilizzare nomi differenti per importare le stesse cose, e questa non è una buona cosa.

Reviewer 2: ```suggestion
Quindi i membri del team possono utilizzare nomi differenti per importare le stesse cose, e questa non è una buona cosa.
```

> Solitamente, per evitare questo problema, e mantenere il codice consistente, ci si pone come regole che le variabili importate debbano corrispondere ai nomi dei file, esempio:

Reviewer: ```suggestion
Solitamente, per evitare questo problema, e mantenere il codice consistente, ci si pone come regola che le variabili importate debbano corrispondere ai nomi dei file, esempio:
```

...
```

Ancora, alcuni team ritengono questo uno svantaggio piuttosto serio dei default export. Quindi preferiscono utilizzare i named exports. Anche se viene esportata solamente una cosa, questa viene esportata con un nome, senza `default`.

Questo rende il re-export (vedi sotto) più semplice.

## Re-export

La sintassi di "re-export", `export ... from ...` ci consente di importare cose ed esportarle immediatamente (eventualmente con un altro nome), come nell'esempio:

```js
export {sayHi} from './say.js'; // re-export sayHi

export {default as User} from './user.js'; // re-export default
```

Perché questo dovrebbe essere necessario? Vediamolo con un esempio pratico.

> Immaginiamo di scrivere un "package" (pacchetto): una cartella contenente molti moduli, con alcune funzionalità esportate esternamente (strumenti come NPM ci consentono di pubblicare e distribuire questi pacchetti, ma non li utilizzeremo), e molti moduli sono semplicemente "helpers", per uso interno in altri moduli del package.

Reviewer 2: ```suggestion
Immaginiamo di scrivere un "package" (pacchetto): una cartella contenente molti moduli, con alcune funzionalità esportate esternamente (strumenti come NPM ci consentono di pubblicare e distribuire questi pacchetti, ma non li utilizzeremo), e con molti moduli che sono semplicemente "helpers", per uso interno in altri moduli del package.
```

        ...
```

Vorremmo poi esporre le funzionalità del package con un singolo entry point (punto di ingresso).

In altre parole, la persona che volesse utilizzare il nostro package, dovrebbe importare solamente il "main file" `auth/index.js`.

Come vediamo qui:

```js
import {login, logout} from 'auth/index.js'
```

Il "main file", `auth/index.js` esporta tutte le funzionalità che vogliamo fornire con il nostro package.

> L'idea è che gli esterni, gli altri programmatori che utilizzano il nostro package, non dovrebbero preoccuparsi della struttura interna, alla ricerca dei file all'interno del nostro package. Esportiamo solamente ciò che è necessario in `auth/index.js` e teniamo il resto nascosto da occhi indiscreti.

Reviewer 2: ```suggestion
L'idea è che gli esterni, gli altri programmatori che utilizzano il nostro package, non debbano preoccuparsi della struttura interna e o di cercare files tra le cartelle in esso contenute. Esportiamo solamente ciò che è necessario in `auth/index.js` e teniamo il resto nascosto da occhi indiscreti.
```

> Poiché le funzionalità di export sono sparpagliate nel package, possiamo importarle in `auth/index.js` ed esportarle da li:

Reviewer: ```suggestion
Poiché le funzionalità di export sono sparpagliate nel package, possiamo importarle in `auth/index.js` ed esportarle da lì:
```

```js
// 📁 auth/index.js

> // import login/logout e li esportiamo immediatamente

Reviewer 2: ```suggestion
// importiamo login/logout e li esportiamo immediatamente
```

import {login, logout} from './helpers.js';
export {login, logout};

> // import default come User e lo esportiamo

Reviewer 2: ```suggestion
// importiamo default come User e lo esportiamo
```

}
```

Potremmo incontrare due problemi:

1. `export User from './user.js'` non funziona. Genererebbe un errore sintattico.

    Per ri-esportare il default export, dobbiamo scrivere `export {default as User}`, come nell'esempio sopra.

2. `export * from './user.js'` ri-esporta solamente i named exports, ma ignora quelli di default.

>     Nel caso in cui volessimo ri-esportare sia i named che i defaul export, allora abbiamo bisogno di due istruzioni:

Reviewer: ```suggestion
    Nel caso in cui volessimo ri-esportare sia i named che i defaul export, allora avremmo bisogno di due istruzioni:
```

    ```js
    export * from './user.js'; // per ri-esportare i named exports
    export {default} from './user.js'; // per ri-esportare i default export
    ```

> Queste stranezza nella ri-esportazione di un default export, è una delle ragioni del perché alcuni sviluppatori non amano i default export, ma preferiscono quelli di default.

Reviewer: ```suggestion
Queste stranezze nella ri-esportazione di un default export sono tra le ragioni del per cui alcuni sviluppatori non amano i default export, ma preferiscono quelli di default.
```

## Riepilogo

Vediamo tutti i tipi di `export` che abbiamo studiato in questo articolo e nei precedenti.

Puoi controllarli tu stesso, leggendoli e provando a ricordarne il significato:

- Prima della dichiarazione di una classe/funzione/..:
  - `export [default] class/function/variable ...`
- Export indipendente:
  - `export {x [as y], ...}`.
- Re-export:
  - `export {x [as y], ...} from "module"`
  - `export * from "module"` (non ri-esporta i default export).
  - `export {default [as y]} from "module"` (ri-esporta i default export).

Import:

- Named exports da un modulo:
  - `import {x [as y], ...} from "module"`
- Default export:  
  - `import x from "module"`
  - `import {default as x} from "module"`
- Tutto:
  - `import * as obj from "module"`
- Importa il modulo (il suo codice viene eseguito), ma non lo assegna ad alcuna variabile:
  - `import "module"`

> Possiamo inserire le istruzioni di `import/export` in cima o in cosa allo script, non ha importanza.

Reviewer: ```suggestion
Possiamo inserire le istruzioni di `import/export` in cima o in coda allo script, non ha importanza.
```
